src/txmempool.cpp

```diff
@@ -1144,17 +1144,25 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
 
     unsigned nTxnRemoved = 0;
     CFeeRate maxFeeRateRemoved(0);
-    while (!mapTx.empty() && DynamicMemoryUsage() > sizelimit) {
+    while (!mapTx.empty()) {
         indexed_transaction_set::index<descendant_score>::type::iterator it = mapTx.get<descendant_score>().begin();
 
+        // Keep trimming as long as memory is above the maximum.
+        // Also, unless min relay feerate allows it, skim away everything paying <=0 fees.
+        const bool keep_trimming{DynamicMemoryUsage() > sizelimit ||
+            (m_min_relay_feerate.GetFeePerK() > 0 && it->GetModFeesWithDescendants() <= 0)};
+        if (!keep_trimming) break;
+
         // We set the new mempool min fee to the feerate of the removed set, plus the
         // "minimum reasonable fee rate" (ie some value under which we consider txn
         // to have 0 fee). This way, we don't allow txn to enter mempool with feerate
         // equal to txn which were removed with no block in between.
         CFeeRate removed(it->GetModFeesWithDescendants(), it->GetSizeWithDescendants());
-        removed += m_incremental_relay_feerate;
-        trackPackageRemoved(removed);
-        maxFeeRateRemoved = std::max(maxFeeRateRemoved, removed);
+        if (removed >= m_min_relay_feerate) {
```

Comment:
True, it's helpful but not strictly necessary here. I think I'll just take this out of this PR and move it to #27018.

Reply:
I've removed this commit. Going to open followup PR for it shortly

Reply:
Ah, fuzzer crash was because I had removed the "evict everything paying 0 fees in TrimToSize()` commit without removing that invariant check. I've added that back along with documentation for why it's relevant here. Sorry for the flip flop.